### PR TITLE
585 Highway Consensus Protocol Documentation

### PR DESCRIPTION
**Pull request by ACStoneCL**
### Related links

[Document Highway Consensus Protocol #585](https://github.com/casper-network/docs/issues/585)

### Changes

High-level description of the Highway Consensus Protocol with information extracted from the whitepaper and previous documentation.

### Notes

Ran locally without issue.

### CODE REVIEW

**Comment by ACStoneCL** [comment]
@afck I extracted the information for this from previous documentation attempts and the original whitepaper - plus the Litepaper and supporting docs provided by @ipopescu.

I figured it may be easier to review/revamp this when you get a chance rather than try to schedule an info dump meeting given your other commitments.

File: source/docs/casper/design/highway.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
+# The Highway Consensus Protocol
+
+## What is Consensus?
+
+Consensus is the backbone of any distributed network. The decentralized nature of a blockchain requires a method through which disparate entities can come to an agreement on one immutable truth. This involves determining validity of transactions, resolving address conflicts, and finalizing blocks to be added to the network. A consensus protocol is a set of mechanisms and rules within the distributed network with which all actors must comply.
```
Comment (afck) [inline source/docs/casper/design/highway.md:5]: By "address conflicts" you mean something like someone trying to spend 10 CSPR twice from an address (i.e. an account) that has only 10 CSPR?
Maybe just "resolving conflicts"?

Not sure about the formulation "adding blocks to the network". I'd rather say they are added _to_ the chain _by_ the network.

File: source/docs/casper/design/highway.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+These rules outline the type of messages sent over the network, when they are sent and how to process them. Within the context of a blockchain, the consensus protocol decides which blocks are added to the chain, and the order by which they are added. This determines the state of the distributed ledger and ensures that all nodes agree on that state.
+The consensus mechanism will determine how a blockchain addresses the following two main characteristics:
```
Comment (afck) [inline source/docs/casper/design/highway.md:7]: Not sure about "addressing characteristics". Maybe "meeting requirements" or "satisfying properties"?

File: source/docs/casper/design/highway.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+*   **Safety** - All honest nodes eventually agree on the final value. The system is setup in a way where no two honest nodes will report two different blocks at the same height of the blockchain.
+
+*   **Liveness** - The system continues running and adds new block to the chain indefinitely.
```
Comment (afck) [inline source/docs/casper/design/highway.md:11]:
```suggestion
*   **Liveness** - The system continues running and adds new blocks to the chain indefinitely.
```

**Review by afck — APPROVED** [review]

File: source/docs/casper/design/highway.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+These validators run a Byzantine Fault Tolerant protocol, currently Highway, which requires a partially synchronous network. The system will continue to function, so long as the amount of faulty or dishonest nodes does not exceed one-third of the total number of nodes in the network. Nodes that are not faulty are *honest* nodes. In most cases, the system can assume that more than two-thirds of all nodes will actively collaborate to achieve consensus. Therefore, stronger-than-average finality guarantees occur during periods where all nodes are acting honestly. A block's fault tolerance increases beyond one-third as the protocol continues. If all validators are honest, it approaches 100%.
```
Comment (ipopescu) [inline source/docs/casper/design/highway.md:21]: If it's ok to keep the BFT mention above, then this line would need to be updated:
```suggestion
These validators run a Casper network that will continue to function so long as the amount of faulty or dishonest nodes does not exceed one-third of the total number of nodes in the network. Nodes that are not faulty are *honest* nodes. In most cases, the system can assume that more than two-thirds of all nodes will actively collaborate to achieve consensus. Therefore, stronger-than-average finality guarantees occur during periods when all nodes are acting honestly. A block's fault tolerance increases beyond one-third as the protocol continues. If all validators are honest, it approaches 100%.
```

File: source/docs/casper/design/highway.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+The consensus mechanism will determine how a blockchain meets the following requirements:
+
+*   **Safety** - All honest nodes eventually agree on the final value. The system is setup in a way where no two honest nodes will report two different blocks at the same height of the blockchain.
+
+*   **Liveness** - The system continues running and adds new blocks to the chain indefinitely.
+
+Casper networks use a consensus protocol known as *Highway*.
+
+## How does the Casper Network use Highway?
```
Comment (ipopescu) [inline source/docs/casper/design/highway.md:17]:
```suggestion
### How does the Casper Network use Highway?
```

File: source/docs/casper/design/highway.md [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+These validators run a Byzantine Fault Tolerant protocol, currently Highway, which requires a partially synchronous network. The system will continue to function, so long as the amount of faulty or dishonest nodes does not exceed one-third of the total number of nodes in the network. Nodes that are not faulty are *honest* nodes. In most cases, the system can assume that more than two-thirds of all nodes will actively collaborate to achieve consensus. Therefore, stronger-than-average finality guarantees occur during periods where all nodes are acting honestly. A block's fault tolerance increases beyond one-third as the protocol continues. If all validators are honest, it approaches 100%.
+
+## Dynamic Round Length
+
+Within the Highway protocol, the length of a round is determined dynamically to ensure a suitable amount of time for nodes to gossip all messages through several round trips with honest validators. This ensures that the system maintains liveness by ensuring that all messages are properly gossiped, while also maintaining a timely addition of blocks to the chain.
+
+## Eras
```
Comment (ipopescu) [inline source/docs/casper/design/highway.md:27]:
```suggestion
### Eras
```

File: source/docs/casper/design/highway.md [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+In any given era, node operators will bid to become validators that will participate in the consensus mechanism for the era after next (`N`+2). Each time slot within the era will also determine a lead validator. The lead validator proposes new blocks to be added to the chain, which are then gossiped among the network's nodes. These messages show an implicit preference for the lead validator's block due to the GHOST (Greedy Heaviest Observed Sub-Tree) rule. Once this process reaches critical mass, with a sufficient interconnected pattern of messages, it becomes impossible to switch to another block. The selected block is then considered finalized and added to the chain.
+
+The final block of an era is a *switch block* and forms the initial state of the next era. A new Highway instance begins with the new era, using information contained within the *switch block* and a new potential set of validators. More details on the auction process to determine an era's validators can be found within the [Consensus Economics](/economics/consensus/) page.
+
+## Finality
```
Comment (ipopescu) [inline source/docs/casper/design/highway.md:41]:
```suggestion
### Finality
```

File: source/docs/casper/design/highway.md [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+These validators run a Byzantine Fault Tolerant protocol, currently Highway, which requires a partially synchronous network. The system will continue to function, so long as the amount of faulty or dishonest nodes does not exceed one-third of the total number of nodes in the network. Nodes that are not faulty are *honest* nodes. In most cases, the system can assume that more than two-thirds of all nodes will actively collaborate to achieve consensus. Therefore, stronger-than-average finality guarantees occur during periods where all nodes are acting honestly. A block's fault tolerance increases beyond one-third as the protocol continues. If all validators are honest, it approaches 100%.
+
+## Dynamic Round Length
+
+Within the Highway protocol, the length of a round is determined dynamically to ensure a suitable amount of time for nodes to gossip all messages through several round trips with honest validators. This ensures that the system maintains liveness by ensuring that all messages are properly gossiped, while also maintaining a timely addition of blocks to the chain.
```
Comment (ipopescu) [inline source/docs/casper/design/highway.md:25]:
```suggestion
Within the Highway protocol, the length of a round is determined dynamically to ensure a suitable amount of time for nodes to gossip all messages through several round trips with honest validators. This ensures that the system maintains liveness by ensuring that all messages are properly gossiped while maintaining a timely addition of blocks to the chain.
```

File: source/docs/casper/design/highway.md [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+*   **Liveness** - The system continues running and adds new blocks to the chain indefinitely.
+
+Casper networks use a consensus protocol known as *Highway*.
+
+## How does the Casper Network use Highway?
+
+The Casper Network is a [Proof-of-Stake](/glossary/P/#proof-of-stake) network, in which the on-chain auction contract determines validators that participate in Highway. The protocol uses a decentralized network of [nodes](/glossary/N/#node), either bonded or unbonded. Nodes that actively participate in the process of consensus must stake CSPR tokens and are known as [Validator Nodes](/glossary/V/#validator). The top 100 bidders are selected through the auction contract every era, to act as validators in the era after the next (Current Era + 2). Nodes with a greater stake in the success of the network have a greater weight in reaching consensus. Highway does not necessitate a Proof-of-Stake method of choosing validators and could theoretically be used alongside a private network with a different model.
```
Comment (ipopescu) [inline source/docs/casper/design/highway.md:19]:
```suggestion
The Casper Network is a [Proof-of-Stake](/glossary/P/#proof-of-stake) network in which the on-chain auction contract determines validators participating in Highway. The protocol uses a decentralized network of [nodes](/glossary/N/#node), either bonded or unbonded. Nodes actively participating in the consensus process must stake CSPR tokens and are known as [Validator Nodes](/glossary/V/#validator). The top 100 bidders are selected through the auction contract every era to act as validators in the era after the next (Current Era + 2). Nodes with a greater stake in the network's success have a greater weight in reaching consensus. Highway does not necessitate a Proof-of-Stake method of choosing validators and could theoretically be used alongside a private network with a different model.
```

File: source/docs/casper/design/highway.md [hunk 10]
```diff
@@ -0,0 +1,47 @@
+# The Highway Consensus Protocol
+
+## What is Consensus?
+
+Consensus is the backbone of any distributed network. The decentralized nature of a blockchain requires a method through which disparate entities can come to an agreement on one immutable truth. This involves determining validity of transactions, resolving conflicts, and finalizing blocks to be added to the chain by the network. A consensus protocol is a set of mechanisms and rules within the distributed network with which all actors must comply.
```
Comment (ipopescu) [inline source/docs/casper/design/highway.md:5]:
```suggestion
Consensus is the backbone of any distributed network. The decentralized nature of a blockchain requires a method through which disparate entities can agree on one immutable truth. This involves determining the validity of transactions, resolving conflicts, and finalizing blocks to be added to the chain by the network. A consensus protocol is a set of mechanisms and rules within the distributed network with which all actors must comply.
```

File: source/docs/casper/design/highway.md [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+These rules outline the type of messages sent over the network, when they are sent and how to process them. Within the context of a blockchain, the consensus protocol decides which blocks are added to the chain by the network, and the order by which they are added. This determines the state of the distributed ledger and ensures that all nodes agree on that state.
```
Comment (ipopescu) [inline source/docs/casper/design/highway.md:7]:
```suggestion
These rules outline the type of messages sent over the network, when they are sent and how to process them. Within the context of a blockchain, the consensus protocol decides which blocks are added to the chain by the network and the order in which they are added. This determines the state of the distributed ledger and ensures that all nodes agree on that state.
```

**Review by ipopescu — APPROVED** [review]
Thank you, @ACStoneCL. I've added small nitpicks from using Grammarly; feel free to accept or deny them. 
Also, I suggested a few additions to improve the flow and tagged you and @afck for a final pass.

**Pull request merged**